File: python/day08.py
from aoc import *
from itertools import combinations


def parse_input(input):
    lines = parse_lines(input)
    size = len(lines)
    antennas = list2grid(lines, lambda x: x != '.')
    return size, antennas

def calc_antinodes(size, pt1, pt2, multis):
    antis = set()
    dist = pt_sub(pt2, pt1)
    for n in multis:
        n_dist = pt_mul(n, dist)
        anti = pt_add(pt2, n_dist)
        if is_inside(size, anti):
            antis.add(anti)
        else:
            break
    return antis

def antinodes(size, antennas, multis):
    acc = set()
    for (pt1, freq1), (pt2, freq2) in combinations(antennas.items(), 2):
        if freq1 == freq2:
            acc |= calc_antinodes(size, pt1, pt2, multis) | calc_antinodes(size, pt2, pt1, multis)
    return len(acc)

def solve(filename):
    data = parse_input(read_input(filename))
    return antinodes(*data, [1]), antinodes(*data, range(100))


solve(8)

File: elixir/day08.exs
defmodule Day08 do
  defp parse_input(filename) do
    import AOC
    lines = filename |> read_input |> parse_lines(:chars)
    size = length(lines)
    antennas = grid_to_point_map(lines, &(&1 != "."))
    {size, antennas}
  end

  defp antinodes({size, antennas}, multis) do
    for {pt1, freq1} <- antennas,
        {pt2, freq2} <- antennas,
        freq1 == freq2,
        pt1 != pt2,
        into: MapSet.new() do
      dist = AOC.pt_sub(pt2, pt1)

      Enum.reduce_while(multis, MapSet.new(), fn n, acc ->
        pt = AOC.pt_add(pt2, AOC.pt_mul(n, dist))

        if AOC.inside?(pt, size) do
          {:cont, MapSet.put(acc, pt)}
        else
          {:halt, acc}
        end
      end)
    end
    |> Enum.reduce(&MapSet.union/2)
    |> Enum.count()
  end

  def solve(filename) do
    data = filename |> parse_input
    {antinodes(data, [1]), antinodes(data, Stream.iterate(0, &(&1 + 1)))}
  end
end

Day08.solve(8)
